conda/recipes/kvikio/meta.yaml

```diff
@@ -64,11 +64,13 @@ requirements:
     - rapids-build-backend >=0.3.0,<0.4.0.dev0
     - scikit-build-core >=0.10.0
     - libkvikio ={{ version }}
+    - aws-sdk-cpp>=1.11.267
```

Comment:
Sorry my ignore here, but why do we need an exact match? Naively, I would have thought that a lower-limit would be more flexible? Do all packages in conda-forge needs to maintain this pinning if they use `aws-sdk-cpp`?

Reply:
First I'll explain the migration logic. Then I'll answer the questions you posed about other options.

### Migrations and why they're needed

The `aws-sdk-cpp` release cadence is very frequent. Some of those releases ([there's a huge number of them](https://github.com/aws/aws-sdk-cpp/tags)) are picked up by conda-forge, but taking the new version requires "rebuilding the world" because `aws-sdk-cpp` is a core C++ dependency of many packages (see `mamba repoquery whoneeds -c conda-forge aws-sdk-cpp`). Updates happen via a conda-forge migration ([tracked here](https://conda-forge.org/status/)) which rebuild everything that depends on `aws-sdk-cpp`.

See the number of recent AWS-related rebuilds of libarrow, for instance:
https://github.com/conda-forge/arrow-cpp-feedstock/commits/main/recipe/meta.yaml

I'll ignore the `aws-crt-cpp` and other rebuilds, and only focus on `aws-sdk-cpp` because that's the dependency we are introducing here.

In the past 12 months, there have been 6 rebuilds of libarrow for new versions of `aws-sdk-cpp` ([1.11.379](https://github.com/conda-forge/arrow-cpp-feedstock/commit/8d5cd0e6e4b7150d99ea0b3c28f968500fdb9ba9), [1.11.329](https://github.com/conda-forge/arrow-cpp-feedstock/commit/1e381317006297d34d53bfda80cb427990a77532), [1.11.267](https://github.com/conda-forge/arrow-cpp-feedstock/commit/2da6505474c2724f24867a2b1434c5c9ad89cd49), [1.11.242](https://github.com/conda-forge/arrow-cpp-feedstock/commit/2fab9347fab92d72d8e8845f864545e83b92ec80), [1.11.210](https://github.com/conda-forge/arrow-cpp-feedstock/commit/e950ac4f208cddef50a79181389cf7f5a24e1fc8), [1.11.182](https://github.com/conda-forge/arrow-cpp-feedstock/commit/1608eaba13cfadc79580b6cc3f1ca7adb5dfffa7)).

Every time that conda-forge migrates to a new version, we must use the _same_ version so that RAPIDS (KvikIO specifically) can be installed alongside the latest conda-forge packages.

Try this:
```
mamba create -n test --dry-run libarrow
```

Note that `aws-sdk-cpp 1.11.379` is included in the environment because there has been a recent rebuild of `libarrow`.

### Can we use a lower limit (or no version pinning)?

If we were to use a lower limit (or no version pinning), kvikio would use the latest `aws-sdk-cpp` when it is built. That would impose a runtime pinning on the latest `aws-sdk-cpp`. However, if a conda-forge migration has not happened yet (or is incomplete), those `kvikio` packages would be incompatible with the latest conda-forge packages of things like libarrow. We need to match the migrators so we lag behind "latest `aws-sdk-cpp`" and instead align with "current version used by conda-forge". Typically conda-forge maintainers will start a migration as soon as a new `aws-sdk-cpp` is released, but that's not guaranteed to happen immediately. Sometimes migrations are [delayed](https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/5913#issuecomment-2119201067) or [skipped](https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/4883#issuecomment-1709403242), which would leave our packages unusable if we had picked up an `aws-sdk-cpp` version that conda-forge did not adopt with a migration.


### So what do we do?

RAPIDS is not built with conda-forge infrastructure, but aims to be compatible with the conda-forge ecosystem. We do not benefit from conda-forge's automation for version migrations but we must track them in order to be compatible. (The conda-forge bots that "rebuild the world" do not rebuild RAPIDS recipes, but maybe we could create such a tool.)

To remain compatible with conda-forge, we need to match the exact pinnings for core packages like `fmt`, `spdlog`, `aws-sdk-cpp`, and others. We have a lot of pain from matching conda-forge migrations for `fmt` and `spdlog` (see [Keith's warning](https://github.com/rapidsai/build-planning/issues/56#issuecomment-2288846888) and [this problem in cuspatial](https://github.com/rapidsai/cuspatial/pull/1453#issuecomment-2335527542) that comes from us being behind on spdlog/fmt migrations)

Our best course of action for now is to:
- pin `aws-sdk-cpp` to the exact version in this YAML file: https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/f291cfdaf8c328337cb8cd1f63c63caceeda8991/recipe/conda_build_config.yaml#L267-L268 (that's a permalink but we should track the latest as it changes)
  - we cannot pin more loosely -- `aws-sdk-cpp` imposes run-exports that pin very tightly, which typically indicates that there is no guarantee of stable ABI across versions
- manually update that version when conda-forge issues a migration
- consider adding tooling to help automate migrations

Reply:
I opened https://github.com/rapidsai/build-planning/issues/100 to think more about conda-forge migration support in RAPIDS.

Reply:
Wow, thanks for the explanation @bdice! I now see why you were hesitant from the get-go 😊  

What if we move to use `libcurl` instead? They take API and ABI stability very serious: https://curl.se/libcurl/features.html#stableapi

Reply:
I was actually thinking of using libcurl from the start since we only need the very basic S3 operations. I don’t think it will be too hard to implement and it will make Azure Blob Storage support straightforward.

Reply:
That might be a better option than `aws-sdk-cpp` if we can use `curl`. It seems like `curl` is already a dependency of some packages we depend on in RAPIDS, so that gives me more confidence in it. See `$ mamba repoquery whoneeds -c conda-forge curl`.

Reply:
Great, I will give it a try